Create balance.py

jrh-ccfe opened the pull request:
This code creates a plasma object from balance.nc files written by SOLPS-ITER

jrh-ccfe committed:
Create balance.py

This code creates a plasma object from balance.nc files written by SOLPS-ITER

jrh-ccfe committed:
Update __init__.py

Updated init.py to include load_solps_from_balance

jrh-ccfe committed:
Update __init__.py

Included load_solps_from_balance

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+
+import os
+import copy
+from numpy import abs,sum,swapaxes,ones,arange
```
It's generally preferable to do `import numpy as np` when you're using `abs` and `sum`, as these overwrite Python's built-in functions. They appear much further down the file too, by which time a reader may have forgotten or missed that they're imported from numpy and wonder why you're passing the `axis` keyword when the Python functions don't support that. Could you please change these calls to `np.sum`, `np.abs` etc for clarity?

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+from cherab.solps.mesh_geometry import SOLPSMesh
+from cherab.solps.solps_plasma import SOLPSSimulation
+
+Q = 1.602E-19
```
Could you use `scipy.constants.elementary_charge` instead please? This is also used internally by Cherab core. The fewer hard coded, slightly different values of e that we have scattered through the code base the better.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+
+
+# Code developed by J. Harrison 9/4/2019
+def load_solps_from_balance(balance_filename, debug=False):
```
You don't currently use the `debug` argument. Either remove it, or add some `if debug:` statements if you want to print out debugging information.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+
+    """
+
+    if not os.path.isdir(balance_filename):
+        RuntimeError("file name must be valid")
```
This block has no effect: an exception will only be raised if you use the `raise` statement. Currently you're just creating an instance of an exception and then not using it.

In any case, this block is redundant. If the file name does not point to a valid netCDF file then line 61 will raise a `FileNotFoundError` anyway. So this block can be removed.

jrh-ccfe replied:
Done - this block has been removed

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    vol = copy.deepcopy(fhandle.variables['vol'].data)
+	
+    # Re-arrange the array dimensions in the way CHERAB expects...
+    cr_x = swapaxes(cr_x,0,2)
+    cr_x = swapaxes(cr_x,0,1)
```
You can accomplish this in a single line. Since you're really moving the first dimension to the end, you can do:
```python
cr_x = np.moveaxis(cr_x, 0, -1)
```
I personally find this version better conveys your actual intent.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    # Re-arrange the array dimensions in the way CHERAB expects...
+    cr_x = swapaxes(cr_x,0,2)
+    cr_x = swapaxes(cr_x,0,1)
+    cr_z = swapaxes(cr_z,0,2)
+    cr_z = swapaxes(cr_z,0,1)
```
Same comment re: moveaxis.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    fhandle = netcdf.netcdf_file(balance_filename,'r')
+	
+    # Load SOLPS mesh geometry
+    cr_x = copy.deepcopy(fhandle.variables['crx'].data)
+    cr_z = copy.deepcopy(fhandle.variables['cry'].data)
+    vol = copy.deepcopy(fhandle.variables['vol'].data)
```
`deepcopy` is unnecessary here: you can use the numpy array's `copy` method instead to create a new array which owns its data. It's neater and faster, and saves an import:
```python
cr_x = fhandle.variables['crx'].data.copy()
```

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    # TODO: add code to load SOLPS velocities and magnetic field from files
+
+    # Load electron species
+    sim._electron_temperature = copy.deepcopy(fhandle.variables['te'].data)/Q
+    sim._electron_density = copy.deepcopy(fhandle.variables['ne'].data)
```
Same comment as above re: deepcopy.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+        charge = int(fhandle.variables['za'].data[i])  # Ionisation/charge
+        species = _popular_species[(zn, am)]
+		
+        # If we only need to populate species_list, there is probably a faster way to do this...		
```
Since the number of species is typically small, you probably don't need to worry about performance. Your intent here is clear: that's more important.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+        # If we only need to populate species_list, there is probably a faster way to do this...		
+        sim.species_list.append(species.symbol + str(charge))
+
+    tmp = copy.deepcopy(fhandle.variables['na'].data)
```
Same comment as above re: deepcopy.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+        sim.species_list.append(species.symbol + str(charge))
+
+    tmp = copy.deepcopy(fhandle.variables['na'].data)
+    tmp = swapaxes(tmp,0,2)
+    tmp = swapaxes(tmp,0,1)
+    sim._species_density = tmp
```
Same comment as above re: moveaxis.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    inside_outside = AxisymmetricMapper(Discrete2DMesh(mesh.vertex_coords, mesh.triangles, inside_outside_data, limit=False))
+    sim._inside_mesh = inside_outside
+	
+    # Load the neutrals data
+    if 'D0' in sim.species_list:
+        for i in arange(len(sim.species_list)):
+            if sim.species_list[i] == 'D0':
+                D0_indx = i
+
+    # Replace the deuterium neutrals density (from the fluid neutrals model by default) with
+    # the values calculated by EIRENE - do the same for other neutrals?
+    if 'dab2' in fhandle.variables.keys():
+        sim.species_density[:,:,D0_indx] = copy.deepcopy(fhandle.variables['dab2'].data[0,:,0:-2])
+        eirene_run = True
+    else:
+        eirene_run = False
```
There are a few issues here:
1. `D0_indx` is only defined if D0 appears in `sim.species_list`. If the balance file somehow has a `dab2` variable but no D0 species then `D0_indx` in line 134 will be undefined and you'll get an `UnboundLocalError` exception raised. For robustness, you could set `D0_indx` to a sentinel value such as `None` beforehand, and only update the neutral density if `D0_indx` is not `None`.
2. Lines 125-129 can be rewritten as: `D0_indx = sim.species_list.index('D0')`. Python has batteries included!
3. Same comment about deepcopy.

You can re-write this section more compactly as the following:
```python
try:
    D0_indx = sim.species_list.index('D0')
except ValueError:  # No 'D0'
    eirene_run = False
    # Could instead raise an error here, if a lack of D0 is a critical error.
else:
    if 'dab2' in fhandle.variables.keys():
        sim.species_density[:, :, D0_indx] = fhandle.variables['dab2'].data[0, :, 0:-2].copy()
        eirene_run = True
    else:
        eirene_run = False
```

I also ran into an error in this part when testing against a balance.nc file I have: 
```python
In [59]: sim = load_solps_from_balance('/home/jlovell/Bolometer/SXD phantoms/Full data/seed_8.nc')       
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-59-de7e0aeae2c8> in <module>
----> 1 sim = load_solps_from_balance('/home/jlovell/Bolometer/SXD phantoms/Full data/seed_8.nc')

~/cherab/cherab_solps/cherab/solps/formats/balance.py in load_solps_from_balance(balance_filename, debug)
    132     # the values calculated by EIRENE - do the same for other neutrals?
    133     if 'dab2' in fhandle.variables.keys():
--> 134         sim.species_density[:,:,D0_indx] = copy.deepcopy(fhandle.variables['dab2'].data[0,:,0:-2])
    135         eirene_run = True
    136     else:

ValueError: could not broadcast input array from shape (34,151) into shape (34,148)
```
Looks like there is a dimension issue happening somewhere.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    # Calculate the total radiated power
+    
+    
+    if eirene_run is True:
```
`if eirene_run:` is sufficient: no need to compare identity to `True`.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+            eirene_potential_loss = 13.6*sum(fhandle.variables['eirene_mc_papl_sna_bal'].data,axis=(0))[1,:,:]*Q/vol
+        
+        # This will be negative (energy sink); take absolute valu
+        sim._total_rad = abs(b2_ploss+(eirene_ecoolrate-eirene_potential_loss))
```
Is it the absolute value we want here? Or do we strictly want to return the negative of the energy loss for the radiated power? Semantically I think the latter, but is there ever the possibility of the energy loss term returned by Eirene being positive (which would correspond to negative radiated power)?

jrh-ccfe replied:
Done, the term in brackets is now multiplied by -1

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+        sim._total_rad = abs(b2_ploss+(eirene_ecoolrate-eirene_potential_loss))
+        
+    else:
+        #print(b2_ploss.shape)
+        #print(cr_x.shape)
+        #print(sum(b2_ploss,axis=0).shape)
```
Please remove commented out code. Or, wrap it in an `if debug:` block and make use of your debug argument.

jrh-ccfe replied:
Done

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+    return sim
+
+
+def load_mesh_from_files(mesh_file_path, debug=False):
+    """
+    Load SOLPS grid description from B2 Eirene output file.
+
+    :param str filepath: full path for B2 eirene mesh description file
+    :param bool debug: flag for displaying textual debugging information.
+    :return: tuple of dictionaries. First is the header information such as the version, label, grid size, etc.
+      Second dictionary has a ndarray for each piece of data found in the file.
+    """
+    _, _, geom_data_dict = load_b2f_file(mesh_file_path, debug=debug)
+
+    cr_x = geom_data_dict['crx']
+    cr_z = geom_data_dict['cry']
+    vol = geom_data_dict['vol']
+
+    # build mesh object
+    return SOLPSMesh(cr_x, cr_z, vol)
```
This is identical to a function with the same name in `raw_simulation_files.py`. Overzealous copy and paste?

jrh-ccfe replied:
Exactly right, this function has been removed

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,192 @@
+        
+        potential_loss = sum(fhandle.variables['b2stel_sna_ion_bal'].data,axis=0)/vol
+        
+        #rad_dens = (-sum(b2stel_she_bal,3)+13.6*Q*b2stel_sna_ion_bal[:,:,0])/vol
```
Again, remove commented out code.

jrh-ccfe replied:
Done

jacklovell reviewed (changes requested):
Thanks for this James! It's a very useful addition. I have a few comments which I've added inline for you to address. I also found an error when trying to read a balance.nc file I have in my home directory: see the comment for lines 125-137.

In addition, there are a couple of general formatting issues. Please use spaces after commas everywhere, to keep consistent with [PEP8](https://www.python.org/dev/peps/pep-0008/) and the rest of the Cherab code base. There's a bit of trailing whitespace too: please remove.

jrh-ccfe committed:
Update balance.py

Changed numpy references

jrh-ccfe committed:
Update balance.py

Q now defined using scipy.constants.elementary_charge

jrh-ccfe committed:
Update balance.py

Tidied up code as per Jack's suggestions

jrh-ccfe committed:
Update balance.py

jrh-ccfe committed:
Update balance.py

Fixed bug where radiated power is negative if SOLPS is run with fluid neutrals

jrh-ccfe committed:
Update balance.py

More code tidying...

jrh-ccfe committed:
Update balance.py

Removed references to copy.deepcopy

jrh-ccfe committed:
Update balance.py

Fixed previously incorrect references to numpy functions

jrh-ccfe committed:
Update balance.py

Added proper handling for case where D0_indx is not set

jrh-ccfe committed:
Fixed indentation error on line 94

jrh-ccfe committed:
Update balance.py

Replaced code that returns D0_indx

jrh-ccfe committed:
Update balance.py

Fixed another indentation issue

jrh-ccfe committed:
Update balance.py

Fixed bug in reading EIRENE atomic density

jrh-ccfe committed:
Create mastu_solps_from_balance.py

Created demo file

jrh-ccfe committed:
Rename mastu_solps_from_balance.py to solps_from_balance.py

Changed file name to a machine agnostic one

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+# under the Licence.
+
+import os
+import copy
```
These imports are no longer needed: please remove them.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+from cherab.core.atomic.elements import hydrogen, deuterium, helium, beryllium, carbon, nitrogen, oxygen, neon, \
+    argon, krypton, xenon
+
+from cherab.solps.eirene import Eirene
```
This import is unused: please remove.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+
+    sim = SOLPSSimulation(mesh)
+    ni = mesh.nx
+    nj = mesh.ny
```
These are unused, and can be removed.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+
+    # Open the file
+    fhandle = netcdf.netcdf_file(balance_filename,'r')
+	
```
You have trailing whitespace here, along with quite a few other places in the file. Please remove it. A quick way to do this is:
```bash
sed -i 's/[[:blank:]]\+$//g' balance.py
```

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+    """
+
+    # Open the file
+    fhandle = netcdf.netcdf_file(balance_filename,'r')
```
Per [PEP-8](https://www.python.org/dev/peps/pep-0008/), please add a space after the comma. This applies to quite a few places throughout the file. Running `pylint balance.py` will point out the others.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+    # TODO: add code to load SOLPS velocities and magnetic field from files
+
+    # Load electron species
+    sim._electron_temperature = fhandle.variables['te'].data.copy()/Q
```
PEP8 again: please put spaces around the `/` operator.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+
+    
+    # Load the neutrals data
+    D0_indx = sim.species_list.index('D0')
```
This will raise a `ValueError` if there is no 'D0' in the species list. Probably safest to wrap this in a try/except block if you want to be able to handle balance.nc files without neutral deuterium:
```python
try:
    D0_indx = sim.species_list.index('D0')
except ValueError:
    D0_indx = None
```
Of course, if missing D0 data is a critical error then there's no point catching this, but then you don't need to do the `if D0_indx is not None` check in line 127.

jacklovell commented on cherab/solps/formats/balance.py:
```diff
@@ -0,0 +1,165 @@
+        potential_loss = np.sum(fhandle.variables['b2stel_sna_ion_bal'].data,axis=0)/vol
+        
+        # Save total radiated power to the simulation object
+        sim._total_rad = 13.6*Q*potential_loss-b2_ploss
```
Please add spaces around the multiplication and subtraction operators here, as well as the division operators in the lines above. Pylint will tell you if you've missed any.

jacklovell commented on demos/solps_from_balance.py:
```diff
@@ -0,0 +1,163 @@
+xl, xu = (0.0, 2.0)
+yl, yu = (-2.0, 2.0)
+
+print('Enter name of balance.nc file:')
+filename = input()
```
I like that users can specify any balance.nc file here in the demo, but you do need to be careful about the assumptions you make. For example, not all runs will have D0, D1 and C0-C6: I have some C-Mod runs with no carbon. I wouldn't worry about modifying the code to catch all these different possibilities, but I would suggest adding a comment here to say that for this demo to run properly the input file must match the assumptions you're making about its contents.

jacklovell commented on demos/solps_from_balance.py:
```diff
@@ -0,0 +1,163 @@
+plt.ylim(yl, yu)
+plt.title("CVII density")
+
+plt.figure()
+plt.imshow(d0_velocity, extent=[xl, xu, yl, yu], origin='lower')
+plt.colorbar()
+plt.xlim(xl, xu)
+plt.ylim(yl, yu)
+plt.title("D0 velocity")
+
+plt.figure()
+plt.imshow(d1_velocity, extent=[xl, xu, yl, yu], origin='lower')
+plt.colorbar()
+plt.xlim(xl, xu)
+plt.ylim(yl, yu)
+plt.title("D1 velocity")
```
The D0 and D1 bulk velocities are not defined, so these plots will fail. Delete this section.

jacklovell reviewed (changes requested):
Thanks for the updated code. I have only a few more stylistic comments, for consistency with the rest of the code base, and a few smaller suggestions. Once these have been addressed I think this will be ready for merging.

jrh-ccfe committed:
Update balance.py

Updated to conform to PEP8 standards

jrh-ccfe committed:
Update solps_from_balance.py

Removed velocity profile plots, added disclaimer that D and C species are assumed

jacklovell reviewed (approved):
Thanks for the updates James. Happy to merge this now.
